TensorLib/NonEmptyList.lean

```diff
@@ -0,0 +1,118 @@
+import Mathlib.Tactic
+
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jean-Baptiste Tristan, Paul Govereau, Sean McLaughlin
+-/
+namespace TensorLib
+
+variable {a: Type}
+
+structure NonEmptyList (a: Type) where
+  hd: a
+  tl: List a
+  deriving BEq, Inhabited
+
+namespace NonEmptyList
+
+def toList (x: NonEmptyList a): List a := x.hd :: x.tl
+
+-- Use the nice list bracket concrete syntax
+instance NelRepr [Repr a] : Repr (NonEmptyList a) where
+  reprPrec x n := reprPrec x.toList n
+
+def length (x: NonEmptyList a): Nat := 1 + x.tl.length
+
+@[simp]
+def append (x : NonEmptyList a) (y : NonEmptyList a) : NonEmptyList a :=
+  NonEmptyList.mk x.hd (x.tl ++ y.toList)
+
```

Comment:
Is it time to start using docstrings to document all this code?

Reply:
Not sure. What do you have in mind?

Reply:
I think he means this: https://leanprover-community.github.io/contribute/doc.html. ?

Reply:
Probably we should setup doc-gen4 in the TenCert parent repo, and generate all docs there ?